atom/browser/api/atom_api_session.cc

```diff
@@ -214,54 +214,82 @@ std::map<uint32_t, v8::Global<v8::Object>> g_sessions;
 void RunCallbackInUI(const base::Callback<void()>& callback) {
   base::PostTaskWithTraits(FROM_HERE, {BrowserThread::UI}, callback);
 }
+
 template <typename... T>
 void RunCallbackInUI(const base::Callback<void(T...)>& callback, T... result) {
   base::PostTaskWithTraits(FROM_HERE, {BrowserThread::UI},
                            base::BindOnce(callback, result...));
 }
 
+void RejectPromise(atom::util::Promise promise, int error) {
+  std::string errmsg =
+      "Failed to size failed with error number: " + std::to_string(error);
+  promise.RejectWithErrorMessage(errmsg);
+}
+
+void RejectPromiseInUI(atom::util::Promise promise, int error) {
+  base::PostTaskWithTraits(
+      FROM_HERE, {BrowserThread::UI},
+      base::BindOnce(RejectPromise, std::move(promise), error));
+}
+
+template <typename T>
+void ResolvePromise(atom::util::Promise promise, T result) {
+  promise.Resolve(result);
+}
+
+template <typename T>
+void ResolvePromiseInUI(atom::util::Promise promise, T result) {
+  base::PostTaskWithTraits(
+      FROM_HERE, {BrowserThread::UI},
+      base::BindOnce(ResolvePromise<T>, std::move(promise), result));
+}
+
 // Callback of HttpCache::GetBackend.
 void OnGetBackend(disk_cache::Backend** backend_ptr,
                   Session::CacheAction action,
-                  const net::CompletionCallback& callback,
+                  const atom::util::CopyablePromise& promise,
                   int result) {
   if (result != net::OK) {
-    RunCallbackInUI(callback, result);
+    RejectPromiseInUI(promise.GetPromise(), result);
   } else if (backend_ptr && *backend_ptr) {
     if (action == Session::CacheAction::CLEAR) {
       (*backend_ptr)
-          ->DoomAllEntries(base::Bind(&RunCallbackInUI<int>, callback));
+          ->DoomAllEntries(
+              base::BindOnce(&ResolvePromiseInUI<int>, promise.GetPromise()));
     } else if (action == Session::CacheAction::STATS) {
       base::StringPairs stats;
       (*backend_ptr)->GetStats(&stats);
       for (const auto& stat : stats) {
         if (stat.first == "Current size") {
           int current_size;
           base::StringToInt(stat.second, &current_size);
-          RunCallbackInUI(callback, current_size);
+          ResolvePromiseInUI(promise.GetPromise(), current_size);
           break;
         }
       }
     }
   } else {
-    RunCallbackInUI<int>(callback, net::ERR_FAILED);
+    RejectPromiseInUI(promise.GetPromise(), net::ERR_FAILED);
```

Comment:
This condition is not required, as the top level `result != net::OK` will cover it.